src/api/ZoweExplorerZosmfApi.ts

```diff
@@ -36,26 +41,178 @@ class ZosmfApiCommon implements ZoweExplorerApi.ICommon {
+                    return {status: "active", session: validateSession};
+                } else {
+                    return {status: "inactive", session: undefined};
+                }
+            } catch (err) {
+                throw new Error(err);
+            }
+        } else {
+            return {status: "unverified", session: undefined};
+        }
+    }
 
-            if (sessionStatus) {
-                return "active";
-            } else {
-                return "inactive";
+    public async getValidSession(serviceProfile: IProfileLoaded,
+                                 profileName: string,
+                                 prompt?: boolean): Promise<Session | null> {
+        // Retrieve baseProfile
+        const baseProfile = DefaultProfileManager.getInstance().getDefaultProfile("base");
+
+        // If user exists in serviceProfile, use serviceProfile to login because it has precedence over baseProfile
+        // If no user & no baseProfile, use serviceProfile as default
+        // If no user & no token, use serviceProfile as default
+        if (serviceProfile.profile.user || !baseProfile || (!serviceProfile.profile.user && baseProfile && !baseProfile.profile.tokenValue)) {
+            if (prompt) {
+                // Select for prompting only fields which are not defined
+                const schemaArray = [];
+                if (!serviceProfile.profile.user && (!baseProfile || (baseProfile && !baseProfile.profile.user))) {
+                    if (!baseProfile || (baseProfile && !baseProfile.profile.tokenValue)) {
+                        schemaArray.push("user");
+                    }
+                }
+                if (!serviceProfile.profile.password && (!baseProfile || (baseProfile && !baseProfile.profile.password))) {
+                    schemaArray.push("password");
+                }
+                if (!serviceProfile.profile.host &&  (!baseProfile || (baseProfile && !baseProfile.profile.host))) {
+                    schemaArray.push("host");
+                    if (!serviceProfile.profile.port &&  (!baseProfile || (baseProfile && !baseProfile.profile.port)))
+                    { schemaArray.push("port"); }
+                }
+
+                try {
+                    const newDetails = await collectProfileDetails(schemaArray);
+                    for (const detail of schemaArray) { serviceProfile.profile[detail] = newDetails[detail]; }
+                } catch (error) {
+                    // tslint:disable:no-magic-numbers
+                    if (error.mDetails && error.mDetails.errorCode === 401) {
+                        if (globals.ISTHEIA) {
+                            vscode.window.showErrorMessage(error.message);
+                            this.getValidSession(serviceProfile, serviceProfile.name, true);
+                        } else {
+                            vscode.window.showErrorMessage(error.message, "Check Credentials").then(async (selection) => {
```

Comment:
This is a really large method in the api file.  The api file usually only handles calls to the api/cli and should never have lines like this making `vscode.window..` or other vscode calls within it.  Is there another way we could handle this method within ZE files and leave the api file strictly for api calls?

Reply:
Let's be sure to discuss this topic in the next squad stand-up. Thanks, @JillieBeanSim , for posting.

Reply:
I guess the question is does it need to be added in the API file? Maybe we can just move it to Profiles.ts or ./profiles/utils.ts? Just thinking out loud.

Reply:
because we still have the getStatus API if ever extenders wanted to ping their URL.

Reply:
I put it here to get rid of a circular dependency. We can try to put it in profiles but it will be calling getSession in the API file and then the circular dependency will return...

Reply:
Let me check this.